js_modules/dagster-ui/packages/ui-core/src/insights/ActivityChart.tsx

```diff
@@ -0,0 +1,196 @@
+import {
+  Body,
+  BodyLarge,
+  BodySmall,
+  Box,
+  Colors,
+  Mono,
+  Popover,
+  Spinner,
+  Subheading,
+} from '@dagster-io/ui-components';
+import {memo} from 'react';
+
+import {TooltipCard} from './InsightsChartShared';
+import {formatDuration} from '../ui/formatDuration';
+import {numberFormatterWithMaxFractionDigits} from '../ui/formatters';
+import {useFormatDateTime} from '../ui/useFormatDateTime';
+import styles from './css/ActivityChart.module.css';
+
+type MetricsDialogData<T> = {
+  after: number;
+  before: number;
+  metric: T;
+  unit: string;
+};
+
+export type ActivityChartDayData = {
+  date: number;
+  hourlyValues: Array<number | null>;
+};
+
+export type ActivityChartData = {
+  max: number | null;
+  dataByDay: ActivityChartDayData[];
+  header: string;
+  color: string;
+  hoverColor: string;
+};
+
+interface Props<T> {
+  metrics: ActivityChartData;
+  unit: string;
+  loading: boolean;
+  openMetricDialog?: (data: MetricsDialogData<T>) => void;
+  metric: T;
+}
+
+const InnerActivityChart = <T,>(props: Props<T>) => {
+  const {metrics, unit, loading, openMetricDialog, metric} = props;
+  const {header, color, dataByDay, max, hoverColor} = metrics;
+  return (
+    <div className={styles.activityChartContainer}>
+      <Box
+        flex={{direction: 'row', alignItems: 'center', justifyContent: 'space-between'}}
+        padding={{bottom: 12}}
+      >
+        <BodyLarge>{header}</BodyLarge>
+        {loading ? <Spinner purpose="body-text" /> : null}
+      </Box>
+      <div className={styles.activityChart}>
+        {dataByDay.map((dayData) => (
+          <ActivityChartRow
+            key={dayData.date}
+            date={dayData.date}
+            hourlyValues={dayData.hourlyValues}
+            max={max}
+            color={color}
+            hoverColor={hoverColor}
+            unit={unit}
+            onClick={openMetricDialog}
+            metric={metric}
+          />
+        ))}
+        <div className={styles.activityChartRow}>
+          <div />
+          <div className={styles.activityChartBottomLegend}>
+            <BodySmall>12AM</BodySmall>
+            <BodySmall>6AM</BodySmall>
+            <BodySmall>12PM</BodySmall>
+            <BodySmall>6PM</BodySmall>
+            <BodySmall>12AM</BodySmall>
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+export const ActivityChart = memo(InnerActivityChart) as typeof InnerActivityChart;
```

Comment:
Pretty annoying to have to do this, but there's no clear way to make `memo` work with generic types, as far as I can tell.

Reply:
yeah thats why I was doing this before 😬